Check operator can be pulled before building catalog

jewzaam opened the pull request:
Can't build the catalog if the referenced operator image is not available.

jewzaam committed:
Check operator can be pulled before building catalog

lisa reviewed (commented):
@jewzaam I'm not sure I agree with this. There is value in doing development locally to see what the output of the catalog is, and if the image must be in the external image repository beforehand we lose that value.

I think this shouldn't be coupled to the operators' presence in the upstream registry. Is there a use case I'm missing?

jewzaam commented:
@lisa good point, what I was trying to catch was actually local development.  to test the catalog in cluster the image has to be available.  but that's only if it's being applied.  maybe there is a better time to do this check?  Maybe just a target that checks for dependent images and call it if/when you care?

lisa commented:
@jewzaam We have to think about the possible workflows. One is local development and one is a production-ready deployment. It's only the latter workflow that we need to be belt and bracers careful that dependencies are lined up. I favour a "UNIX-like" pipeline where the step prior to this repo's involvement is to validate that the relevant operators all have their current images in their respective registries.

That looks like:

1. New operator is built
2. Logic to ensure the build is successful (success meaning: build completes, image is in the registry, etc)
3. Create a new catalog with this repository.
4. Some process updates clusters to have the most recent Subscription and/or CatalogSource data

Flow is handed off by whatever logic is performing these steps only once the previous step is successful. Where are we doing this workflow?

jewzaam commented:
@lisa we aren't yet and not sure where we will.  I'll switch this out to add an independent target that you can opt to use.  For now it will be useful to use before trying to deploy the operator to a cluster, confirm the dependencies actually exist.  We dont' have to use it in that workflow you outlined.

jewzaam committed:
Move check dependent images to independent target: check-operator-images

jewzaam commented:
Actually, it can be run before "push".  At that point you're pushing this image out, any dependent images must exist for it to be usable.

jewzaam committed:
Moved check-operator-images as dependency for push target

If you are pushing osd-operators-registry you expect the latest operator images to be pullable.

lisa commented on README.md:
```diff
@@ -67,7 +67,7 @@ The following are some of the `make` targets are included:
 - catalog - updates the `catalog-manifests/` if there are any updates
 - build - build the container image
-- push - pushes the container image
+- push - pushes the container image (after verifying operator's currentCSV images exist)
```
You should add a note about the `check-operator-images` target in this list too.

jewzaam replied:
It's a dependency of `push` and we don't call out all the dependencies.  Is there a use case where someone would want to use it that didn't require a push?

lisa replied:
I didn't realize we omit the dependencies. Carry on. ;)

lisa commented on Makefile:
```diff
@@ -122,6 +122,13 @@ catalog: operator-source
 	done
 
+.PHONY: check-operator-images
+check-operator-images: operator-source
```
It would be cool if we could iterate over `$(OPERATORS)` list here, but I think that requires a change to all the operators' Makefiles to change their `env` targets to spit out the image registry URI.

jewzaam replied:
It could be passed explicitly from `make env`. Would rather deal with separately if that's something we want because of dependency on the operator changes.

lisa replied:
It's almost worth it, especially if the image registry or organizational bits are different ... 🤔  hrmm...

jewzaam replied:
https://github.com/openshift/dedicated-admin-operator/pull/27

jewzaam replied:
I would like to merge this PR before adding in update to use `OPERATOR_IMAGE_URI`.  https://github.com/openshift/osd-operators-registry/issues/8

lisa replied:
Sounds reasonable.

/lgtm
